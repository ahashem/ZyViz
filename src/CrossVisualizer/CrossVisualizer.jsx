import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'speedux';
import {
  Spin,
} from 'antd';

import './style.scss';
import module from './CrossVisualizer.module';


class CrossVisualizer extends Component {
  static propTypes = {
    actions: PropTypes.shape({
    }),
    state: PropTypes.shape({
      loading: PropTypes.bool,
      error: PropTypes.bool,
      errorMsg: PropTypes.string,
    })
  };

  static defaultProps = {
    actions: {},
    state: {
      columns: [],
      allColumns: [],
      campaignsCount: 0,
      current: 1,
    },
    history: {},
    location: {},
  };

  state = {

  };

  componentDidMount() {

  }

  componentDidUpdate(prevProps) {

  }

  render() {

    return (
      <div>
        <div>All Charts will render here</div>
        <Spin />
      </div>
    );
  }
}

export default connect(CrossVisualizer, module);
