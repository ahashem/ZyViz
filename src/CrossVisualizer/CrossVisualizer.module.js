import { createModule } from 'speedux';
import { OrdersAPI } from '../utils/API';

export default createModule('CrossVisualizer', {

  state: {
    loading: false,
    error: false,
    errorMsg: '',
    chartsData: {},
  },

  actions: {
    /**
     * this action get all Charts data
     */
    * getOrdersData() {
      yield {
        loading: true,
      };

      const allData = yield OrdersAPI.getOrders().catch(e => e);

      if (allData instanceof Error) {
        const { errorMsg } = JSON.parse(allData.message);
        yield {
          loading: false,
          error: true,
          errorMsg,
        };
      }

      const chartsData = allData.data;

      if (chartsData) {
        yield {
          loading: false,
          chartsData,
        };
      }
    },
  },

  handlers: {

  },
});
