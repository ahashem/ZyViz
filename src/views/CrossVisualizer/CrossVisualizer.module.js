import { createModule } from 'speedux';
import { OrdersAPI } from '../../utils/API';

export default createModule('CrossVisualizer', {

  state: {
    loading: false,
    error: false,
    errorMsg: '',
    ordersData: [],
  },

  actions: {
    /**
     * this action get all Charts data
     */
    * getOrdersData() {
      yield {
        loading: true,
      };

      const ordersData = yield OrdersAPI.getOrders().catch(e => e);

      if (ordersData instanceof Error) {
        const { errorMsg } = JSON.parse(ordersData.message);
        yield {
          loading: false,
          error: true,
          errorMsg,
        };
      } else {
        yield {
          loading: false,
          ordersData,
        };
      }
    },
  },

  handlers: {

  },
});
