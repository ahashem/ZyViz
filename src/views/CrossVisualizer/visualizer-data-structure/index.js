import NormalizeData from './data-normalizer';
import GenerateDimensions from './dimensions-manager';

export { NormalizeData, GenerateDimensions };
