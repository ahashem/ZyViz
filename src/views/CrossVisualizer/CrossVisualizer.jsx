import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'speedux';
import { Spin } from 'antd';
import crossfilter from '../../components/crossfilter/crossfilter';

import module from './CrossVisualizer.module';
import OrdersCountPie from './OrdersCountPie/OrdersCountPie';

import './CrossVisualizer.scss';


class CrossVisualizer extends Component {
  static propTypes = {
    actions: PropTypes.shape({
      getOrdersData: PropTypes.func
    }).isRequired,
    state: PropTypes.shape({
      loading: PropTypes.bool,
      error: PropTypes.bool,
      errorMsg: PropTypes.string,
      ordersData: PropTypes.arrayOf(PropTypes.shape({
        userid: PropTypes.string,
        orderid: PropTypes.number,
        orderAmount: PropTypes.string,
        orderdate: PropTypes.string,
        paymentMethod: PropTypes.string,
        branch: PropTypes.string,
        deliveryArea: PropTypes.string,
      }))
    }).isRequired
  };

  state = {};

  componentDidMount() {
    const { actions } = this.props;
    actions.getOrdersData();
  }

  componentDidUpdate(prevProps) {
  }

  normalizeData = (ordersData) => {
    const currencyParser = (value) => value && parseFloat(value.replace(/[$,]+/g, ''));

    // Normalize data for Crossfilter usage!
    ordersData.forEach(order => {
      order.orderAmount = currencyParser(order.orderAmount);
    });
    return ordersData;
  };

    render() {

    const { state } = this.props;
    const { loading, ordersData } = state;

    const orders = (ordersData && ordersData.length) ? crossfilter(this.normalizeData(ordersData)) : null;

    return (
      <div>
        {loading ? (<Spin/>)
          : (
            <React.Fragment>
              <h3>Orders by <br /><strong>payment method</strong></h3>
              <OrdersCountPie
                orders={orders}
              />
            </React.Fragment>
          )}
      </div>
    );
  }
}

export default connect(CrossVisualizer, module);
