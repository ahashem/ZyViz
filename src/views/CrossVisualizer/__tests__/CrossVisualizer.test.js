import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import CrossVisualizer from '../CrossVisualizer';

const mockOrders = [
  {
    'userid': '94a16117-9db3-4f25-abdf-1c3ef6578097',
    'orderid': 36,
    'orderAmount': '$51.89',
    'orderdate': '2015-03-16T19:46:12+00:00',
    'paymentMethod': 'Cash',
    'branch': 'Branch D',
    'deliveryArea': 'Mirqab'
  },
  {
    'userid': '0aed2c89-8d05-4a0c-b5dd-35c24829020e',
    'orderid': 37,
    'orderAmount': '$22.77',
    'orderdate': '2015-11-06T12:33:56+00:00',
    'paymentMethod': 'KNET',
    'branch': 'Branch A',
    'deliveryArea': 'Funaitees'
  },
];


test('should render children and match snapshot', () => {
  const crossVisualizer = {
    loading: false,
    error: false,
    errorMes: '',
    ordersData: mockOrders,
  };
  const mockStore = configureStore();
  const store = mockStore({ crossVisualizer });

  const rendered = shallow(<CrossVisualizer store={store}/>);

  expect(rendered).toMatchSnapshot();
});

describe('<CrossVisualizer /> Component', () => {
  describe('UI during loading', () => {
    let wrapper;

    beforeAll(() => {
      const crossVisualizer = {
        loading: true,
      };
      const mockStore = configureStore();
      const store = mockStore({ crossVisualizer });

      wrapper = mount(<CrossVisualizer store={store}/>);
    });

    test('it should render loading spinner while loading', () => {
      const spinner = wrapper.find('[data-test="loading"]');
      expect(spinner.exists()).toBe(true);
    });

  });

});
