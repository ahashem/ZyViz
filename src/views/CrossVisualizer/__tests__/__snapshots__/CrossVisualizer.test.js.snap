// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CrossVisualizer /> Component should render children and match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(CrossVisualizer)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "actions": Object {
        "getOrdersData": [Function],
        "updateFilters": [Function],
      },
      "state": Object {
        "error": false,
        "errorMsg": "",
        "loading": false,
        "ordersData": Object {
          "add": [Function],
          "all": [Function],
          "allFiltered": [Function],
          "dimension": [Function],
          "groupAll": [Function],
          "isElementFiltered": [Function],
          "onChange": [Function],
          "remove": [Function],
          "size": [Function],
        },
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "actions": Object {
          "getOrdersData": [Function],
          "updateFilters": [Function],
        },
        "state": Object {
          "error": false,
          "errorMsg": "",
          "loading": false,
          "ordersData": Object {
            "add": [Function],
            "all": [Function],
            "allFiltered": [Function],
            "dimension": [Function],
            "groupAll": [Function],
            "isElementFiltered": [Function],
            "onChange": [Function],
            "remove": [Function],
            "size": [Function],
          },
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
